substrate/frame/staking/src/pallet/impls.rs

```diff
@@ -977,11 +981,20 @@ impl<T: Config> Pallet<T> {
 	/// wrong.
 	pub fn do_add_nominator(who: &T::AccountId, nominations: Nominations<T>) {
 		if !Nominators::<T>::contains_key(who) {
-			// maybe update sorted list.
-			let _ = T::VoterList::on_insert(who.clone(), Self::weight_of(who))
-				.defensive_unwrap_or_default();
+			// new nominator.
+			Nominators::<T>::insert(who, nominations);
+			<T::EventListeners as OnStakingUpdate<T::AccountId, BalanceOf<T>>>::on_nominator_add(
+				who,
+			);
+		} else {
+			// update nominations.
+			let prev_nominations = Self::nominations(who).unwrap_or_default();
+			Nominators::<T>::insert(who, nominations);
+			<T::EventListeners as OnStakingUpdate<T::AccountId, BalanceOf<T>>>::on_nominator_update(
```

Comment:
Just curious, would it be better to just have one fn `on_nominator_update` instead of three (update, add, remove) and passing old and new nomination as parameter to it? I feel it might be much easier to maintain.

Also a bit confused about `on_nominator_idle` and `on_nominator_remove`? Seems both means nominator stops nominating. I see there is a case where nominator unbonds completely but that should be covered by `on_stake_update` right?

Reply:
> Just curious, would it be better to just have one fn on_nominator_update instead of three (update, add, remove) and passing old and new nomination as parameter to it? I feel it might be much easier to maintain.

That could be done yes, but I think that it's actually easier to read and maintain if we split the code path of a) updating the stake of the ledger and b) update the nominations/validation state. That's why we don't only have the `on_stake_update` to cover all updates in the stake-tracker. So I think there should be a balance, and I'm relatively happy with the current `OnStakingUpdate` trait (although I'm actually suggesting to add a new handler `on_*_idle`).

> Also a bit confused about on_nominator_idle and on_nominator_remove? Seems both means nominator stops nominating. 

The question boils down to why should we split the `OnStakingUpdate::on_staker_remove` (note that `staker` == `nominator` or `validator`) for `on_staker_remove` and `on_staker_idle`. IMO, there are three main reasons why we should split thise:
1. First, and as a quick reminder, this PR will change the way we handle idle (chilled) stakers on the `VoterList` and `TargetList` side, i.e. on the stake-tracker side. The current code, assumes that chilled voter and target nodes will be removed from the `VoterList` and `TargetList`. However, the current `stake-tracker` implementation will **not remove** the nodes from the lists. These two different cases should be "communicated" to the `stake-tracker`. The question now is how should that be communicated? Implicitly or explicitly? I'd say explicitly is better, which I argue about in the two points below.
2. It makes the code more readable at the caller site: it would not as explicit if chilling and removing completely a staker would require calling, e.g, `on_nominator_remove`. This way, we explicitly handle the remove and chill (idle) cases in the caller site, which is easier to read and understand, less prone to errros.
3. It makes the implementation simpler at the `OnStakingUpdate` implementor: there are less checks to make and less reads from the staking interface. The `OnStakingUpdate`  implementor just needs to handle the idle/remove cases.

Reply:
> These two different cases should be "communicated" to the stake-tracker. The question now is how should that be communicated? 

I would say, stake needs to be tracked as long as there is bonded stake. Nominator remove only happens when ledger is killed and that should be consequence of stake dropping to zero (in reality below ExistentialDeposit). Why should there be an explicit way to remove nominator? Isn't that dangerous?

> It makes the code more readable at the caller site

Not sure about this. I see a match statement with 3 cases to add/update a nomination. This could be simplified to following
```rust
pub fn do_add_nominator(who: &T::AccountId, new_nominations: Nominations<T>) {
		let prev_nominations = Self::nominations(who);
		Nominators::<T>::insert(who, new_nominations);
		T::EventListeners::on_nominator_update(who, prev_nominations, Some(new_nominations));
}
```

Probably subjective, but I would also like the explicit trait fn `on_nominator_update(who, prev_nomination: Option<Vec<AccountId>>, new_nominations: Option<Vec<AccountId>>)` where we pass both new and old nominations.

> It makes the implementation simpler at the OnStakingUpdate implementor

Not super convinced about this either. Update fn impl would be more generic but it is essentially doing same thing as combination of add and remove. Less code, less place to update/maintain future changes.